### PR TITLE
[deconz] Added background discovery service

### PR DESCRIPTION
**Pull request by cweitkamp**
- Added background discovery service
- Fixed discovery of `water-leakage` sensor

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

### CODE REVIEW

File: addons/binding/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/HandlerFactory.java [hunk 1]
```diff
@@ -79,12 +89,32 @@ public void unsetHttpClientFactory(HttpClientFactory factory) {
     @Override
     protected @Nullable ThingHandler createHandler(Thing thing) {
         ThingTypeUID thingTypeUID = thing.getThingTypeUID();
-
         if (BRIDGE_TYPE.equals(thingTypeUID)) {
-            return new DeconzBridgeHandler((Bridge) thing, webSocketFactory,
+            DeconzBridgeHandler handler = new DeconzBridgeHandler((Bridge) thing, webSocketFactory,
                     new AsyncHttpClient(httpClientFactory.getCommonHttpClient()));
+            ThingDiscoveryService discoveryService = new ThingDiscoveryService(handler);
```
Comment (davidgraeff) [inline addons/binding/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/HandlerFactory.java]: Is `getServices()` not working? I mean it was meant to be used for such use cases

Comment (cweitkamp) [reply addons/binding/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/HandlerFactory.java]: The problem why it was no working were theses lines of code:

https://github.com/openhab/openhab2-addons/blob/2b9c68e54dc037073beb0ec0d4f1dd5214a3aa60/addons/binding/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/handler/DeconzBridgeHandler.java#L163-L166

The `thingDiscoveryService` has never been assigned to the handler. The implemented solution was the only way I could imagine how to set the missing reference. Is there a better way to do it?

Comment (davidgraeff) [reply addons/binding/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/HandlerFactory.java]: It's a bit of a two way reference situation, I agree. But the pattern with `getServices()` requires this.

What happens is:
* the framework requests the service list via `getServices()`. 
* It starts all those services for the ThingHandler and calls `setThingHandler(handler)`
* And in that method we set the discoveryService reference within the handler. See:

https://github.com/openhab/openhab2-addons/blob/826b0f41e7d3920e712c7c63c6e91d2fe11875d4/addons/binding/org.openhab.binding.deconz/src/main/java/org/openhab/binding/deconz/internal/discovery/ThingDiscoveryService.java#L106

**Comment by cweitkamp** [comment]
Thanks for teaching me something I was not familiar with this procedure. Do you know a similar implementation in another binding? Is it use somewhere else? Or are we dealing with a regression in the framework?

I will try to revert it partially and only add the background discovery. Maybe that will work.

**Comment by davidgraeff** [comment]
> Is it use somewhere else

In all bindings since about 6 months this is the pattern to be used. At least I got corrected for the MQTT binding and Milight binding.

**Comment by cweitkamp** [comment]
I tinkered around a little bit longer with the `getServices()` API but unfortunately I could not get the discovery service to work properly. Some things I observed during my testing:

- component got activated but it did not resolve dependencies on other OSGI bundles (not necessary for this PR but maybe I other cases where you need it right e.g. OpenWeatherMap binding)
- manual discovery did not work via Paper UI which afaik invokes the `startScan()` method
- component was not deactivated properly
  I added the background discovery and a log output in the scheduled method. The log output continued after deactivation of the service.
  A possible reason for this could be the duplicate overload for `deactivate()` method from implementing  the `ThingStatus` interface and extending the abstract `AbstractDiscoveryService` class.

Do not hesitate to ask if you do not understand what I am meaning.

How can we continue on this? Go for a fix in the Core API or pick it up like I suggested?

**Comment by davidgraeff** [comment]
Hm. Extending `AbstractDiscoveryService` is quite common (in other words: a good use-case) for such a `getServices()` service.

Really bad news that it does not work.

> Did not resolve dependencies on other OSGI bundles

How is that even possible. This is a normal OSGI component and all dependencies should be resolved before it can be started. The core must do something really unusual when starting those services.

> manual discovery did not work via Paper UI which afaik invokes the `startScan()` method

Which is strange. If Paper UI offers you the deConz binding as an option for discovery, then the `AbstractDiscoveryService` got successfully registered to the core, including the `startScan()` method.

I would love to core debug that problem but I'm still waiting for a working build setup. In my opinion we should fix core though and not work around those problems as other bindings are also starting to use `getServices()`.

**Comment by cweitkamp** [comment]
> Extending AbstractDiscoveryService is quite common (in other words: a good use-case) for such a getServices() service.

Absolutely. It looks really promising and is a lot easier to understand and to use.

> I would love to core debug that problem

I second and appreciate that. I'd like to help you with that. I will upload my OWM binding test code later this day and reduce this PR to only add background discovery.

**Comment by cweitkamp** [comment]
I think I made a mistake in my test. I called the `activate()` method inside the constructor of my discovery service. This obviously resulted in not resolved dependencies and not working deactivation. The problem is simpler. The services just did not get activated at all (see my playground https://github.com/cweitkamp/openhab2-addons/commit/ce253f9a16e89d9429883e8ca988d44cdd13fa9b#commitcomment-32461727).

**Comment by davidgraeff** [comment]
The part of the core that starts the service:

https://github.com/openhab/openhab-core/blob/master/bundles/org.openhab.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandlerFactory.java#L147

**Comment by cweitkamp** [comment]
> If that assumption is true that the interface name matching doesn't work, we should at least see a call to setThingHandler:

The call happens (see https://github.com/cweitkamp/openhab2-addons/commit/ba83734463454f88dc4e3dc4948ab92a2da0079a).

**Comment by martinvw** [comment]
@cweitkamp I see an approval here, but its title is WIP should we merge or wait?

**Comment by cweitkamp** [comment]
@martinvw Wenn can merge this. We have to do research in the framework to get more information about the real problem. I changed the title of the PR.

**Comment by martinvw** [comment]
Thanks!

**Pull request merged**